### PR TITLE
Pe balance

### PR DESCRIPTION
**Pull request by Paulie14**
- change interface in balance to use local dofs
- update transport.cc assemble routines with DHCellAccessor
- change type Idx to int
- improve a bit code structure of BC assembly in LMH

### CODE REVIEW

File: src/coupling/balance.hh [hunk 1]
```diff
@@ -260,42 +262,34 @@ public:
 			start_source_assembly(idx);
 	}
 
+	
 	/**
-	 * Adds elements into matrix for computing mass.
+	 * Adds elements into matrix for computing (outgoing) flux.
```
Comment (jbrezmorf) [inline src/coupling/balance.hh]: for computing mass.
... document shape of matrix and vector for every kind of the add_* routine.

File: src/coupling/balance.hh [hunk 2]
```diff
@@ -260,42 +262,34 @@ public:
...
 	 */
-	void add_mass_matrix_values(unsigned int quantity_idx,
-			unsigned int region_idx,
-			const std::vector<LongIdx> &dof_indices,
-			const std::vector<double> &values);
+	void add_mass_values(unsigned int quantity_idx,
+			const DHCellAccessor &dh_cell,
+			const LocDofVec &loc_dof_indices,
+			const std::vector<double> &mat_values,
+			const double &vec_value);
```
Comment (jbrezmorf) [inline src/coupling/balance.hh]: vec_value could be rather passed by value.

File: src/coupling/balance.hh [hunk 3]
```diff
@@ -260,42 +262,34 @@ public:
...
-     * @param value         Value to be added.
-     */
-    void add_mass_vec_value(unsigned int quantity_idx,
-            unsigned int region_idx,
-            double value);
+	void add_flux_values(unsigned int quantity_idx,
+			const DHCellSide &side,
+			const LocDofVec &loc_dof_indices,
+			const std::vector<double> &mat_values,
+			const double &vec_value);
```
Comment (jbrezmorf) [inline src/coupling/balance.hh]: vec_value could be rather passed by value.

File: src/coupling/balance.hh [hunk 4]
```diff
@@ -307,20 +301,10 @@ public:
 	 */
 	void add_source_values(unsigned int quantity_idx,
 			unsigned int region_idx,
-			const std::vector<LongIdx> &loc_dof_indices,
+			const LocDofVec &loc_dof_indices,
 			const std::vector<double> &mat_values,
             const std::vector<double> &vec_values);
```
Comment (jbrezmorf) [inline src/coupling/balance.hh]: why we have vector of vec_values here?

File: src/flow/assembly_lmh.hh [hunk 5]
```diff
@@ -235,139 +232,144 @@ protected:
...
+            // side-edge (flux-lambda) terms
+            loc_system_.add_value(loc_side_dofs[sidx], loc_edge_dofs[sidx], 1.0);
+            loc_system_.add_value(loc_edge_dofs[sidx], loc_side_dofs[sidx], 1.0);
+        }
+    }
+
+
+    void assemble_side_bc(const DHCellSide& side, double cross_section, bool use_dirichlet_switch)
+    {
+        const unsigned int sidx = side.side_idx();
```
Comment (jbrezmorf) [inline src/flow/assembly_lmh.hh]: I hope this doesn't change as the diff is a mess.

File: src/flow/assembly_lmh.hh [hunk 6]
```diff
@@ -235,139 +232,144 @@ protected:
...
+
+            double bc_pressure = ad_->bc_pressure.value(b_ele.centre(), b_ele);
+            double bc_switch_pressure = ad_->bc_switch_pressure.value(b_ele.centre(), b_ele);
+            double bc_flux = -ad_->bc_flux.value(b_ele.centre(), b_ele);
+            double bc_sigma = ad_->bc_robin_sigma.value(b_ele.centre(), b_ele);
+
+            double solution_head = ad_->p_edge_solution[loc_schur_.row_dofs[sidx]];
+
+            // Force Robin type during the first iteration of the unsteady case.
+            if (solution_head > bc_switch_pressure  || ad_->force_bc_switch) {
```
Comment (jbrezmorf) [inline src/flow/assembly_lmh.hh]: 'force_bc_switch' rename to 'force_no_neuman_bc' ?

File: src/flow/assembly_lmh.hh (hunk 5 above)
Comment (Paulie14) [reply src/flow/assembly_lmh.hh]: I did this with intention not to do any changes in the code logic, but to separate the switch part of the BC code and to reduce the tabs of all the "embeddings".

File: src/flow/assembly_lmh.hh (hunk 6 above)
Comment (Paulie14) [reply src/flow/assembly_lmh.hh]: Agreed.

**Pull request merged**